### PR TITLE
JAVA-1899 Support Virtual Tables

### PR DESCRIPTION
**Pull request by GregBestland**
Java1899 for 3.x

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: This should be executed conditionnally, we don't need to run this query if the target type is e.g. `FUNCTION`.

I don't think we can get targeted schema notifications -- `targetType == KEYSPACE or TABLE` -- for virtual elements (but we need to double-check that, I haven't looked at the server code). If that is the case, then all virtual metadata (including tables and columns) only needs to be queried on a full schema refresh, we would include them in the `isSchemaOrKeyspace` block above.

File: driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java]: Nit: rename `isVirtual` to `virtual`

File: driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java]: Nit: space between equal sign

File: driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java]: Would be preferable for this constructor should reuse the new one (i.e. `this(name, durableWrites, replication, false)`).

Should also mark this constructor as deprecated as it looks like existing uses of it were updated to use the new one.

File: driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java]: Nit: Add space between line 80 and 81

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: nit: space between `!=`

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java:780]: It would be nice to not duplicate a lot of this code from V3Parser, but refactoring it so it could be reused would probably be as much code, so i'm fine with this as is.

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: Nit: `virtualcolumns` -> `virtualColumns`

File: driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java]: This is currently not passing in columns, so it looks like all virtual tables have no columns.

**Review by tolbertam — CHANGES_REQUESTED** [review]
Looks good for the most part, had a few nits, and there is one issue in that columns are missing from `TableMetadata`.  This could also use a simple integration test that targets cassandra version 4.0+ that simply ensures that keyspaces/tables/columns are present for virtual tables.

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: You could make these protected in `V3SchemaParser` instead of redefining them.

File: driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java:433]: Maybe we should add a `TableMetadata.isVirtual()`, even though it will just be a trivial shortcut to `getKeyspace().isVirtual()`.

File: driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java]: So a virtual table's metadata never has columns? Shouldn't we expose the contents of `system_virtual_schema.columns` here?

**Review by olim7t — CHANGES_REQUESTED** [review]

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (GregBestland) [reply driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java:780]: That was my thought as well.

File: driver-core/src/test/java/com/datastax/driver/core/VirtualTableMetadataTest.java
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/VirtualTableMetadataTest.java:30]: Looks good, just a few suggestions:

1. Inconsistent spacing between operators in some cases (i.e. no space between `=`, `>=` in some places.
2. Would be good to add comments inline for each type validated to indicate why their values are that way, i.e.:

    ``` 
    // Keyspace name should be set, marked as virtual, and have a clients table.
    // All other values should be defaulted since they are not defined in schema tables.
    ```

File: driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java]: Looks like `default )` was accidentally put here.

File: driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java]: nit: spaces between commas.

File: driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/KeyspaceMetadata.java]: Should mark this constructor as deprecated.

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: nit: I see `virtualcolumns` was renamed `virtualColumns` for the class member, but the constructor parameter is still `virtualcolumns`.

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: nit: should be indented another space to match the other arguments.

**Review by tolbertam — COMMENTED** [review]
Looking good :+1:, one small suggestion and some nits.  You mentioned out-of-band something about updating `describe` for virtual metadata, so will hold off on approving until that's sorted.

**Comment by GregBestland** [comment]
Should be everything Andy let me know if I missed anything

**Review by tolbertam — APPROVED** [review]
Looks good! :+1:

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: You need curly braces around this block, currently only the first statement is executed conditionally.

File: driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java]: Nit: the javadoc should say table,  not keyspace.

File: driver-core/src/test/java/com/datastax/driver/core/VirtualTableMetadataTest.java
Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/VirtualTableMetadataTest.java]: Nit: this is duplicated at line 48, I would get rid of this line.

**Review by olim7t — CHANGES_REQUESTED** [review]

**Comment by olim7t** [comment]
LGTM, last round of minor/trivial comments and this is good to go.

File: driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java]: In several places the formatting rules are not respected, please run a global format on this PR before merging, applying the rules for driver 3.x [here](https://github.com/datastax/java-driver/blob/3.x/CONTRIBUTING.md#editor-configuration).

File: driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/TableMetadata.java]: I know this is extremely unlikely but we could theoretically produce invalid CQL here if the table contains weird column names as in ` create table t1 ("pk */" int primary key)`; this could also happen if some table options contain such forbidden sequences, for example in the `comments` field.
One simple idea to fix this would be to use single-line comments (`--` or `//`) and force `formatted` to false.
EDIT: nevermind, with virtual tables this should never happen. OK to ignore this comment.

**Review by adutra — DISMISSED** [review]

File: driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java
Comment (GregBestland) [reply driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java]: Good catch

**Comment by GregBestland** [comment]
Everything should be addressed

**Review by olim7t — APPROVED** [review]

**Pull request merged**
